server/repository/src/db_diesel/changelog/changelog.rs

```diff
@@ -464,6 +471,52 @@ fn create_filtered_outgoing_sync_query(
     query
 }
 
+/// Runs some DB operation with a fully locked `changelog` table.
+/// This only applies for for Postgres and does nothing for Sqlite.
+///
+/// Motivation:
+/// When querying changelog entries, ongoing transactions might continue adding changelog entries
+/// to the queried range of changelogs.
+/// This is because Postgres has Read Committed isolation level (instead of Serialized in Sqlite).
+/// However, we assume that there will be no new changelog entries in the queried range in the
+/// future, e.g. when updating the cursor position.
+///
+/// For example, a changelog may contain [1, 3, 4, 5] while another (slow) tx is about to commit a
+/// changelog row with cursor = 2.
+/// We need to wait for this changelog 2 to be added before doing the changelogs() query, otherwise
+/// we might update the latest changelog cursor to 5 and the changelog with cursor = 2 will be left
+/// unhandled when continuing from the latest cursor position.
+///
+/// Locking the changelog table will wait for ongoing writers and will prevent new writers while
+/// reading the changelog.
+fn with_locked_changelog_table<T, F>(
+    connection: &StorageConnection,
+    f: F,
+) -> Result<T, RepositoryError>
+where
+    F: FnOnce(&mut LockedConnection) -> Result<T, RepositoryError>,
+{
+    if cfg!(feature = "postgres") {
+        use diesel::connection::SimpleConnection;
+        let result = connection.transaction_sync_etc(
+            |con| {
+                let mut locked_con = con.lock();
+                locked_con
+                    .connection()
+                    .batch_execute("LOCK TABLE ONLY changelog IN ACCESS EXCLUSIVE MODE")?;
```

Comment:
A little bit unsure about impact on performance, especially once we get central omSupply server used by lots of sites, while still wanting 'atomic' updates (invoice and invoice_lines to appear in changelog in the same transaction). 

I think ideally would be great if database could set the serials at the end of transaction.

We are actually planning to relocate all trigger insertions of changelog to a manual insert_changelog, maybe that will give use more flexibility, and we would be able to do some optimisations.

Reply:
yes if we manage the cursor on our own we maybe could do better? However, we probably still have to lock our cursor counter for the whole tx which might not help much...

At least the table lock is currently held quite briefly because we only load a handful of changelogs at a time.

But yes some benchmark would be good. Maybe in combination with the test that James suggested then we can test it against a new method.

Reply:
But the processor would only block sync briefly. And blocking the processor during sync is ok, isn't it?

Reply:
Sync (integration) would also block all other UI operations, which would typically be fine, but could be problematic, like if we do sync priority, which allows for UI/UX to be opened without fully initialised (for example temperature data to sync after UI/UX is open for operations), when temperature logs are integrating, UI/UX would be blocked. I think this change should be ok for now, I do think we can do some optimisations in changelogs are inserted via repo. 

I didn't want to get too deep into this, but couldn't resist:

* Change logs are inserted straight away if not within transaction
* If within transaction insert them but also keep record of which ids (serials) have been inserted for which transaction `txid_current()`. When transaction finishes, increment id of those changelogs.

Reply:
Don't think I 100% understand what you mean with your last two points, probably better to have an offline discussion.
Think doing it manually would be quite hard (e.g. handle cases where the server may crash). IMHO the problem can better be solved tackling the long-lasting integration transaction, e.g. making it save to integrate records in batches. (just about doing a PR for integrating in batches during initialization which also brings good speedups for pg and sqlite...)